rpc_jobs/build_gating_venv.yml

```diff
@@ -10,7 +10,73 @@
+          // run in docker to ensure all the required apt packages are available
+          container = docker.build env.BUILD_TAG.toLowerCase()
+        }
+        // Venv is build within docker as it requires extra packages
+        container.inside{
+          sh """#!/bin/bash -xeu
+
+          # Install virtualenv if required
+          if [[ ! -d ".venv" ]]; then
+            requirements="virtualenv==15.1.0"
+            pip install -U "\${requirements}" \
+              || pip install --isolated -U "\${requirements}"
+
+            if which scl
+            then
+              # redhat/centos
+              source /opt/rh/python27/enable
+              virtualenv --no-pip --no-setuptools --no-wheel --python=/opt/rh/python27/root/usr/bin/python .venv
+              # hack the selinux module into the venv
+              cp -r /usr/lib64/python2.6/site-packages/selinux .venv/lib64/python2.7/site-packages/ ||:
+            else
+              virtualenv --no-pip --no-setuptools --no-wheel .venv
+            fi
+          fi
+
+          # Install Pip
+          set +xeu
+          source .venv/bin/activate
+          set -xeu
+
+          # UG-613 change TMPDIR to directory with more space
+          export TMPDIR="/var/lib/jenkins/tmp"
+
+          # If the pip version we're using is not the same as the constraint then replace it
+          PIP_TARGET="\$(awk -F= '/^pip==/ {print \$3}' rpc-gating/constraints.txt)"
+          VENV_PYTHON=".venv/bin/python"
+          VENV_PIP=".venv/bin/pip"
+          if [[ "\$(\${VENV_PIP} --version)" != "pip \${PIP_TARGET}"* ]]; then
+            # Install a known version of pip, setuptools, and wheel in the venv
+            CURL_CMD="curl --silent --show-error --retry 5"
+            OUTPUT_FILE="get-pip.py"
+            \${CURL_CMD} https://bootstrap.pypa.io/get-pip.py > \${OUTPUT_FILE} \
+              || \${CURL_CMD} https://raw.githubusercontent.com/pypa/get-pip/master/get-pip.py > \${OUTPUT_FILE}
+            GETPIP_OPTIONS="pip setuptools wheel --constraint rpc-gating/constraints.txt"
+            \${VENV_PYTHON} \${OUTPUT_FILE} \${GETPIP_OPTIONS} \
+              || \${VENV_PYTHON} \${OUTPUT_FILE} --isolated \${GETPIP_OPTIONS}
+          fi
+
+          # Install rpc-gating requirements
+          PIP_OPTIONS="-c rpc-gating/constraints.txt -r rpc-gating/requirements.txt"
+          \${VENV_PIP} install \${PIP_OPTIONS} \
+            || \${VENV_PIP} install --isolated \${PIP_OPTIONS}
+
+          # Install ansible roles
+          mkdir -p rpc-gating/playbooks/roles
+          ansible-galaxy install -r rpc-gating/role_requirements.yml -p rpc-gating/playbooks/roles
+          """
+        } // container
+
+        // venv is pushed to rpc repo from outside docker container, as binding
+        // credentials files does not work within docker (used for ssh key)
         withCredentials(artifact_build.get_rpc_repo_creds()) {
           sh """#!/bin/bash -xeu
```

Comment:
The upload should be locked to prevent two jobs trying to upload the same venv at the same time, potentially corrupting the artifact.

Reply:
Never mind - I see that the job does not allow concurrency.